### ISSUE TITLE
Why 'Restore Deleted Messages' not working?

### ISSUE THREAD

**Issue by JGVfh** [tomer8007/whatsapp-web-incognito#183]
Hi everybody!

Why is the 'Restore Deleted Messages' feature not working? 

It is no longer functioning on Chrome or Firefox."

**Comment by tomer8007** [comment]
I will try to see...

**Comment by vivekwarde** [comment]
@tomer8007 Any update on this? It randomly works sometimes and sometimes not. This extension has also stopped working:  https://chromewebstore.google.com/detail/wa-web-plus-by-elbruz-tec/ekcgkejcjdcmonfpmnljobemcbpnkamh

**Comment by tomer8007** [comment]
@vivekwarde After some debugging I see that actually message decryption works fine, and detection of deleted messages is also working, but in the end the deleted message is still not shown correctly. I will debug more.

**Comment by esauvisky** [comment]
> @vivekwarde After some debugging I see that actually message decryption works fine, and detection of deleted messages is also working, but in the end the deleted message is still not shown correctly. I will debug more.

If you could give me some basic guidance on where to look, I’d be happy to help you troubleshoot the issue.

**Comment by tomer8007** [comment]
@esauvisky Enable debugging with `WADebugMode = true` to see the incoming and outgoing messages,  look in `checkForMessageDeletionNode`, `onDeletionMessageBlocked`, and the "Blocking incoming REVOKE message node" printing line.

Basically it looks like either the code does not block the REVOKE message correctly, WhatsApp has used another way to mark deletion, or my test case wasn't good.

**Comment by esauvisky** [comment]
@tomer8007 Are you sure the message decryption is working fine? Because I'm running into lots of "Wrong counter in decryption", is that expected? 

Here, it looks like `decryptNoisePacket` returns null, causing the `.after` hook fall back to returning the original `messageEvent`, but I don't know how to tell if these particular frames that fail decryption are relevant.

**Comment by tomer8007** [comment]
@esauvisky Yes, also got these, and that's also why I initially thought that this is the problem, but after some debugging I concluded that those are not relevant for the problem.

I would breakpoint at the functions I listed above, then send a message and delete it (and also enable debug mode). That's the easiest way to see if the extension detects both messages.

**Comment by vivekwarde** [comment]
@tomer8007  have u tried debugging with breakpoints? please keep us updated

**Comment by tomer8007** [comment]
@esauvisky You can see it `checkForMessageDeletionNode`

**Comment by tomer8007** [comment]
> Yes, also got these, and that's also why I initially thought that this is the problem, but after some debugging I concluded that those are not relevant for the problem.

I was wrong, the noise packet decryption is indeed the problem, it's just that sometimes WhatsApp servers send a different kind of handshake pattern in the noise protocol so it seems working.

which explains why

>  It randomly works sometimes and sometimes not

**Comment by tomer8007** [comment]
Anyway, I figured it out.
It's a statistical race problem in the Noise frames decryption logic that is (mostly) caused by a new kind of server hello message that WhatsApp is doing, which makes my code miss a frame, leading the decryption counters to be out of sync.

@esauvisky @vivekwarde @JGVfh This should be fixed in version 2.2.0 (https://github.com/tomer8007/whatsapp-web-incognito/commit/1ca5692660a0d6b17a402669d7cceafbf0dd34c8). I would appreciate if you can test it.

**Comment by esauvisky** [comment]
@tomer8007 Jesus christ, I'm glad I didn't waste more than a couple hours yesterday 🤣

Works ok here, nice job!

**Comment by tomer8007** [comment]
@esauvisky This was really fast :-)

**Comment by tik3ii** [comment]
Working now, even after closing and reopening it!

**Comment by rock82rajesh** [comment]
Hello Sir, I tested it, and at first, it worked fine for up to 10 minutes. However, now it only works continuously for up to 5 minutes. After that, it stops showing deleted messages, and I have to refresh it for it to start working again, but even then, it only works for another 5 minutes.

**Comment by tomer8007** [comment]
@rock82rajesh Are you using version 2.2.1? I had a similar problem in version 2.2.0, and then I fixed it. Pleaes note that at the moment version 2.2.1 is not at the store yet.

**Comment by tomer8007** [comment]
@rock82rajesh Version 2.2..1 is live in the store now. Can you please test it?

**Comment by rock82rajesh** [comment]
@tomer8007 yes sir i tested it for about half n hour and its working fantastic, thanks for fixing it.
